# End-to-End FinTech Payment & Customer Analytics Platform

## Project Overview

This project demonstrates an end-to-end analytics workflow using:
- Python (Pandas)
- MySQL
- SQL
- Power BI

Main tasks:
- data cleaning
- staging table creation
- data warehouse modeling
- SQL analytics
- KPI reporting
- dashboard development

In [67]:
# !/usr/bin/env python3
import pandas as pd
import matplotlib.pyplot as plt
from sqlalchemy import create_engine

In [68]:
# create SQLAlchemy engine to connect to MySQL
engine = create_engine("mysql+pymysql://root:root1234@localhost:3306/fintech_analytics_dw")

In [ ]:
# Load data from CSV files into pandas DataFrames
df_accounts = pd.read_csv('')
df_customers = pd.read_csv('/Users/dineshkumarmuthusamy/Desktop/End-to-End Fintech Payment & Customer Analytics Platform/data_raw/DimCustomer.csv')
df_customers_usa = pd.read_csv('/Users/dineshkumarmuthusamy/Desktop/End-to-End Fintech Payment & Customer Analytics Platform/data_raw/DimCustomerUSA.csv')
df_products = pd.read_csv('/Users/dineshkumarmuthusamy/Desktop/End-to-End Fintech Payment & Customer Analytics Platform/data_raw/Dimproduct.txt', sep=',')
df_products_categories = pd.read_csv('/Users/dineshkumarmuthusamy/Desktop/End-to-End Fintech Payment & Customer Analytics Platform/data_raw/DimProductCategory.csv')
df_products_subcategories = pd.read_csv('/Users/dineshkumarmuthusamy/Desktop/End-to-End Fintech Payment & Customer Analytics Platform/data_raw/DimProductSubCategory.csv')
df_transactions = pd.read_csv('/Users/dineshkumarmuthusamy/Desktop/End-to-End Fintech Payment & Customer Analytics Platform/data_raw/FactTransaction.csv')

In [70]:
# Check the structure of the DataFrame
df_accounts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   AccountID       193 non-null    int64  
 1   CustomerID      193 non-null    int64  
 2   AccountType     193 non-null    object 
 3   OpenDate        193 non-null    object 
 4   ClosedDate      107 non-null    object 
 5   Status          193 non-null    object 
 6   RegistrationID  193 non-null    int64  
 7   Balance         193 non-null    float64
dtypes: float64(1), int64(3), object(4)
memory usage: 12.2+ KB


In [71]:
# Check for missing values in the DataFrame 
df_accounts.isnull().sum()

AccountID          0
CustomerID         0
AccountType        0
OpenDate           0
ClosedDate        86
Status             0
RegistrationID     0
Balance            0
dtype: int64

In [72]:
# Convert date columns to datetime format
df_accounts['OpenDate'] = pd.to_datetime(df_accounts['OpenDate'])

df_accounts['ClosedDate'] = pd.to_datetime(
    df_accounts['ClosedDate'],
    errors='coerce'
)

In [73]:
# Clean and standardize categorical columns
df_accounts['AccountType'] = (
    df_accounts['AccountType']
    .str.strip()
    .str.title()
)

df_accounts['Status'] = (
    df_accounts['Status']
    .str.strip()
    .str.title()
)

In [74]:
# Check for duplicates in the DataFrame
df_accounts.duplicated().sum()

np.int64(0)

In [75]:
# Check for duplicates in the 'AccountID' column
df_accounts['AccountID'].duplicated().sum()

np.int64(0)

Credit Accounts
Negative balances may be VALID. example AccountID: 1 and Balance : -2781.87


In [76]:
# Calculate the age of each account in days
df_accounts['AccountAgeDays'] = (pd.Timestamp.now() - df_accounts['OpenDate']).dt.days


In [77]:
# churn_rate
df_accounts['churn_flag'] = df_accounts['Status'].map({'Open': 0, 'Closed': 1})

In [78]:
# negative balance flag 
df_accounts['NegativeBalanceFlag'] = (df_accounts['Balance'] < 0).astype(int)

In [79]:
# account type distribution
df_accounts['AccountType'].value_counts()

AccountType
Checking    70
Savings     66
Credit      57
Name: count, dtype: int64

In [80]:
# Load the cleaned DataFrame into the MySQL database
df_accounts_clean = df_accounts.copy()
df_accounts_clean.to_sql("stg_accounts", engine, if_exists="replace", index=False)

193

In [81]:
# verify
pd.read_sql("SELECT * FROM stg_accounts LIMIT 5;", engine)

,AccountID,CustomerID,AccountType,OpenDate,ClosedDate,Status,RegistrationID,Balance,AccountAgeDays,churn_flag,NegativeBalanceFlag
0,1,1,Credit,2023-10-31,2024-10-03,Closed,3,-2781.87,930,1,1
1,2,2,Checking,2023-11-08,NaT,Open,1,23903.23,922,0,0
2,3,3,Checking,2020-11-13,NaT,Open,3,42851.83,2012,0,0
3,4,4,Checking,2024-07-01,2024-11-15,Closed,2,41777.99,686,1,0
4,5,5,Savings,2025-05-19,NaT,Open,2,28041.50,364,0,0


In [82]:
# Check for missing values in the customers DataFrame
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   CustomerID  100 non-null    int64 
 1   FullName    100 non-null    object
 2   DOB         100 non-null    object
 3   Gender      100 non-null    object
 4   Region      100 non-null    object
 5   Email       100 non-null    object
 6   Status      100 non-null    object
 7   JoinDate    100 non-null    object
dtypes: int64(1), object(7)
memory usage: 6.4+ KB


In [83]:
df_customers.isnull().sum()

CustomerID    0
FullName      0
DOB           0
Gender        0
Region        0
Email         0
Status        0
JoinDate      0
dtype: int64

In [84]:
# Convert 'DOB' and 'JoinDate' to datetime format
df_customers['DOB'] = pd.to_datetime(
    df_customers['DOB'],
    dayfirst=True
)

df_customers['JoinDate'] = pd.to_datetime(
    df_customers['JoinDate'],
    dayfirst=True
)

In [85]:
# Standardize 'Gender', 'Region', and 'Status' columns
df_customers['Gender'] = df_customers['Gender'].str.strip().str.title()
df_customers['Region'] = df_customers['Region'].str.strip().str.title()
df_customers['Status'] = df_customers['Status'].str.strip().str.title()

In [86]:
# Check for duplicates in the customers DataFrame
df_customers.duplicated().sum()

np.int64(0)

In [87]:
# check 'True/False' check confirmation:
print(df_customers['Email'].str.contains('@').all())

True


In [88]:
# Calculate age from DOB
df_customers['Age'] = (pd.Timestamp.today() - df_customers['DOB']).dt.days // 365

In [89]:
# Create 'IsActive' column based on 'Status'
df_customers['IsActive'] = ( df_customers['Status'] == 'Active').astype(int) 


In [90]:
# validate 'Status' distribution
df_customers['Status'].value_counts()

Status
Suspended    38
Active       33
Inactive     29
Name: count, dtype: int64

In [91]:
# upload to MySQL
df_customer_clean = df_customers.copy()
df_customer_clean.to_sql("stg_customers", engine, if_exists="replace", index=False)

100

In [92]:
# verify
pd.read_sql("SELECT * FROM stg_customers LIMIT 5;", engine)

,CustomerID,FullName,DOB,Gender,Region,Email,Status,JoinDate,Age,IsActive
0,1,Norma Fisher,1977-02-28,Female,Florida,thull@yahoo.com,Active,2023-08-03,49,1
1,2,Susan Wagner,1978-09-21,Female,California,montgomeryjohn@mcgrath.com,Inactive,2021-05-19,47,0
2,3,Stephanie Sutton,1991-09-03,Female,California,ryan93@woods.net,Inactive,2020-08-29,34,0
3,4,Ryan Page,2005-01-04,Male,Alaska,vclayton@cross.com,Inactive,2022-03-02,21,0
4,5,Daniel Pratt,1969-03-28,Male,Colorado,millerluke@hotmail.com,Suspended,2024-09-14,57,0


In [93]:
# Check for missing values in the customers DataFrame
df_customers_usa.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   CustomerID  100 non-null    int64 
 1   FullName    100 non-null    object
 2   DOB         100 non-null    object
 3   Gender      100 non-null    object
 4   Region      100 non-null    object
 5   Email       100 non-null    object
 6   Status      100 non-null    object
 7   JoinDate    100 non-null    object
dtypes: int64(1), object(7)
memory usage: 6.4+ KB


In [94]:
# Check for missing values in the customers DataFrame
df_customers_usa.isnull().sum()

CustomerID    0
FullName      0
DOB           0
Gender        0
Region        0
Email         0
Status        0
JoinDate      0
dtype: int64

In [95]:
# Convert 'DOB' and 'JoinDate' to datetime format
df_customers_usa['DOB'] = pd.to_datetime(
    df_customers_usa['DOB'],
    dayfirst=True
)

df_customers_usa['JoinDate'] = pd.to_datetime(
    df_customers_usa['JoinDate'],
    dayfirst=True
)

In [96]:
# Standardize 'Gender', 'Region', and 'Status' columns
df_customers_usa['Gender'] = df_customers_usa['Gender'].str.strip().str.title()
df_customers_usa['Region'] = df_customers_usa['Region'].str.strip().str.title()
df_customers_usa['Status'] = df_customers_usa['Status'].str.strip().str.title()

In [97]:
# Check for duplicates in the customers DataFrame
df_customers_usa.duplicated().sum()

np.int64(0)

In [98]:
# check 'True/False' check confirmation:
print(df_customers_usa['Email'].str.contains('@').all())

True


In [99]:
# Calculate age from DOB
df_customers_usa['Age'] = (pd.Timestamp.today() - df_customers_usa['DOB']).dt.days // 365

In [100]:
# Create 'IsActive' column based on 'Status'
df_customers_usa['IsActive'] = ( df_customers_usa['Status'] == 'Active').astype(int) 


In [101]:
# upload to MySQL
df_customerusa_clean = df_customers_usa.copy()
df_customerusa_clean.to_sql("stg_customers_usa", engine, if_exists="replace", index=False)

100

In [102]:
# verify
pd.read_sql("SELECT * FROM stg_customers_usa LIMIT 5;", engine)

,CustomerID,FullName,DOB,Gender,Region,Email,Status,JoinDate,Age,IsActive
0,1,Norma Fisher,1977-02-28,Female,Florida,thull@yahoo.com,Active,2023-08-03,49,1
1,2,Susan Wagner,1978-09-21,Female,California,montgomeryjohn@mcgrath.com,Inactive,2021-05-19,47,0
2,3,Stephanie Sutton,1991-09-03,Female,California,ryan93@woods.net,Inactive,2020-08-29,34,0
3,4,Ryan Page,2005-01-04,Male,Alaska,vclayton@cross.com,Inactive,2022-03-02,21,0
4,5,Daniel Pratt,1969-03-28,Male,Colorado,millerluke@hotmail.com,Suspended,2024-09-14,57,0


In [103]:
# Check the structure of the Clean DataFrame
df_products.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27 entries, 0 to 26
Data columns (total 3 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   ProductID             27 non-null     int64 
 1   ProductSubcategoryID  27 non-null     int64 
 2   ProductName           27 non-null     object
dtypes: int64(2), object(1)
memory usage: 780.0+ bytes


In [104]:
# Check for missing values in the products DataFrame
df_products.isnull().sum()

ProductID               0
ProductSubcategoryID    0
ProductName             0
dtype: int64

In [105]:
# Check for duplicates in the products DataFrame
df_products['ProductID'].duplicated().sum()

np.int64(0)

In [106]:
# Standardize 'ProductName' column
df_products['ProductName'] = (
    df_products['ProductName']
    .str.strip()
    .str.upper()
)

In [107]:
# upload to MySQL
df_products_clean = df_products.copy()
df_products_clean.to_sql("stg_products", engine, if_exists="replace", index=False)

27

In [108]:
# verify
pd.read_sql("SELECT * FROM stg_products LIMIT 5;", engine)

,ProductID,ProductSubcategoryID,ProductName
0,1000,100,AAA
1,1001,100,AAB
2,1002,100,AAC
3,1003,101,ABA
4,1004,101,ABB


In [109]:
# datatype check
df_products_categories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 2 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   ProductCategoryID    3 non-null      int64 
 1   ProductCategoryName  3 non-null      object
dtypes: int64(1), object(1)
memory usage: 180.0+ bytes


In [110]:
# null check
df_products_categories.isnull().sum()


ProductCategoryID      0
ProductCategoryName    0
dtype: int64

In [111]:
# duplicate check
df_products_categories['ProductCategoryID'].duplicated().sum()


np.int64(0)

In [112]:
# upload to MySQL
df_products_categories_clean = df_products_categories.copy()
df_products_categories_clean.to_sql("stg_products_categories", engine, if_exists="replace", index=False)

3

In [113]:
# verify
pd.read_sql("SELECT * FROM stg_products_categories LIMIT 5;", engine)

,ProductCategoryID,ProductCategoryName
0,1,A
1,2,B
2,3,C


In [114]:
# datatype check
df_products_subcategories.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 3 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   ProductSubCategoryID    9 non-null      int64 
 1   ProductCategoryID       9 non-null      int64 
 2   ProductSubCategoryName  9 non-null      object
dtypes: int64(2), object(1)
memory usage: 348.0+ bytes


In [115]:
# null check
df_products_subcategories.isnull().sum()


ProductSubCategoryID      0
ProductCategoryID         0
ProductSubCategoryName    0
dtype: int64

In [116]:
# duplicate check
df_products_subcategories['ProductSubCategoryID'].duplicated().sum()


np.int64(0)

In [117]:
# upload to MySQL
df_products_subcategories_clean = df_products_subcategories.copy()
df_products_subcategories_clean.to_sql("stg_products_subcategories", engine, if_exists="replace", index=False)

9

In [118]:
# verify
pd.read_sql("SELECT * FROM stg_products_subcategories LIMIT 5;", engine)

,ProductSubCategoryID,ProductCategoryID,ProductSubCategoryName
0,100,1,AA
1,101,1,AB
2,102,1,AC
3,103,2,BA
4,104,2,BB


In [119]:
# datatype check
df_transactions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   TransactionID       10000 non-null  int64  
 1   AccountID           10000 non-null  int64  
 2   TransactionDate     10000 non-null  object 
 3   TransactionAmount   10000 non-null  float64
 4   TransactionType     10000 non-null  object 
 5   TransactionChannel  10000 non-null  object 
 6   ProductID           10000 non-null  int64  
 7   Status              10000 non-null  object 
dtypes: float64(1), int64(3), object(4)
memory usage: 625.1+ KB


In [120]:
# null check
df_transactions.isnull().sum()

TransactionID         0
AccountID             0
TransactionDate       0
TransactionAmount     0
TransactionType       0
TransactionChannel    0
ProductID             0
Status                0
dtype: int64

In [121]:
# duplicate check
df_transactions.duplicated().sum()


np.int64(0)

In [122]:
# convert transaction date
df_transactions['TransactionDate'] = pd.to_datetime(
    df_transactions['TransactionDate'],
    errors='coerce'
)

In [123]:
# standardize 'TransactionType', 'TransactionChannel', and 'Status' columns
df_transactions['TransactionType'] = (df_transactions['TransactionType'].str.strip().str.title())
df_transactions['TransactionChannel'] = (df_transactions['TransactionChannel'].str.strip().str.title())
df_transactions['Status'] = (df_transactions['Status'].str.strip().str.title())

In [124]:
# transaction amount change to numeric
df_transactions['TransactionAmount'] = pd.to_numeric(df_transactions['TransactionAmount'], errors='coerce')

In [125]:
# unique transaction count
df_transactions['TransactionID'].nunique()

10000

In [126]:
# transaction status distribution
df_transactions['Status'].value_counts()

Status
Success    9533
Failed      467
Name: count, dtype: int64

In [127]:

# upload to MySQL
df_transactions_clean = df_transactions.copy()
df_transactions_clean.to_sql("stg_transactions", engine, if_exists="replace", index=False)

10000

In [128]:
# verify
pd.read_sql("SELECT * FROM stg_transactions LIMIT 5;", engine)

,TransactionID,AccountID,TransactionDate,TransactionAmount,TransactionType,TransactionChannel,ProductID,Status
0,1,162,2023-02-24,10000.00,Credit,Web,1019,Success
1,2,154,2022-08-31,10000.00,Credit,Web,1001,Success
2,3,179,2020-01-03,4883.80,Debit,Web,1024,Success
3,4,138,2023-11-29,-3911.93,Credit,Atm,1021,Success
4,5,159,2024-02-18,-2211.72,Credit,Mobile,1001,Success


In [129]:
# foreign key validation: Ensure all 'AccountID' in transactions exist in accounts
df_transactions['AccountID'].isin(
    df_accounts['AccountID']
).all()

np.True_

# SQL Data Warehouse & Analytics Layer

After completing data cleaning and validation in Pandas,
the cleaned staging tables were uploaded into MySQL.

The next phase focuses on:
- data warehouse structure
- primary and foreign key relationships
- SQL-based business logic
- KPI calculations
- analytical queries
- reporting views for Power BI integration

In [130]:
# verify that all tables are loaded in MySQL 
pd.read_sql("SHOW TABLES;", engine)

,Tables_in_fintech_analytics_dw
0,stg_accounts
1,stg_customers
2,stg_customers_usa
3,stg_products
4,stg_products_categories
5,stg_products_subcategories
6,stg_transactions


In [131]:
# Create dim_customer and load data with transformations
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""
        DROP TABLE IF EXISTS dim_customers;
    """))

    conn.execute(text("""
        CREATE TABLE dim_customers AS
        SELECT * FROM stg_customers;
    """))

In [132]:
# verify 
pd.read_sql("SELECT * FROM dim_customers LIMIT 5;", engine)

,CustomerID,FullName,DOB,Gender,Region,Email,Status,JoinDate,Age,IsActive
0,1,Norma Fisher,1977-02-28,Female,Florida,thull@yahoo.com,Active,2023-08-03,49,1
1,2,Susan Wagner,1978-09-21,Female,California,montgomeryjohn@mcgrath.com,Inactive,2021-05-19,47,0
2,3,Stephanie Sutton,1991-09-03,Female,California,ryan93@woods.net,Inactive,2020-08-29,34,0
3,4,Ryan Page,2005-01-04,Male,Alaska,vclayton@cross.com,Inactive,2022-03-02,21,0
4,5,Daniel Pratt,1969-03-28,Male,Colorado,millerluke@hotmail.com,Suspended,2024-09-14,57,0


In [133]:
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""
        DROP TABLE IF EXISTS dim_customers_usa;
    """))

    conn.execute(text("""
        CREATE TABLE dim_customers_usa AS
        SELECT * FROM stg_customers_usa;
    """))

In [134]:
pd.read_sql("SELECT * FROM dim_customers_usa LIMIT 5;", engine)

,CustomerID,FullName,DOB,Gender,Region,Email,Status,JoinDate,Age,IsActive
0,1,Norma Fisher,1977-02-28,Female,Florida,thull@yahoo.com,Active,2023-08-03,49,1
1,2,Susan Wagner,1978-09-21,Female,California,montgomeryjohn@mcgrath.com,Inactive,2021-05-19,47,0
2,3,Stephanie Sutton,1991-09-03,Female,California,ryan93@woods.net,Inactive,2020-08-29,34,0
3,4,Ryan Page,2005-01-04,Male,Alaska,vclayton@cross.com,Inactive,2022-03-02,21,0
4,5,Daniel Pratt,1969-03-28,Male,Colorado,millerluke@hotmail.com,Suspended,2024-09-14,57,0


In [135]:
# Create dim_accounts and load data
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""
        DROP TABLE IF EXISTS dim_accounts;
    """))

    conn.execute(text("""
        CREATE TABLE dim_accounts AS
        SELECT * FROM stg_accounts;
    """))

In [136]:
# verify
pd.read_sql("SELECT * FROM dim_accounts LIMIT 5;", engine)

,AccountID,CustomerID,AccountType,OpenDate,ClosedDate,Status,RegistrationID,Balance,AccountAgeDays,churn_flag,NegativeBalanceFlag
0,1,1,Credit,2023-10-31,2024-10-03,Closed,3,-2781.87,930,1,1
1,2,2,Checking,2023-11-08,NaT,Open,1,23903.23,922,0,0
2,3,3,Checking,2020-11-13,NaT,Open,3,42851.83,2012,0,0
3,4,4,Checking,2024-07-01,2024-11-15,Closed,2,41777.99,686,1,0
4,5,5,Savings,2025-05-19,NaT,Open,2,28041.50,364,0,0


In [137]:
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""
        DROP TABLE IF EXISTS dim_products;
    """))

    conn.execute(text("""
        CREATE TABLE dim_products AS
        SELECT * FROM stg_products;
    """))
    

In [138]:
# verify
pd.read_sql("SELECT * FROM dim_products LIMIT 5;", engine)

,ProductID,ProductSubcategoryID,ProductName
0,1000,100,AAA
1,1001,100,AAB
2,1002,100,AAC
3,1003,101,ABA
4,1004,101,ABB


In [139]:
# Create dim_products_categories and load data
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""
        DROP TABLE IF EXISTS dim_products_categories;
    """))

    conn.execute(text("""
        CREATE TABLE dim_products_categories AS
        SELECT * FROM stg_products_categories;
    """))

In [140]:
# verify
pd.read_sql("SELECT * FROM dim_products_categories LIMIT 5;", engine)

,ProductCategoryID,ProductCategoryName
0,1,A
1,2,B
2,3,C


In [141]:
# Create dim_products_subcategories and load data
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""
        DROP TABLE IF EXISTS dim_products_subcategories;
    """))

    conn.execute(text("""
        CREATE TABLE dim_products_subcategories AS
        SELECT * FROM stg_products_subcategories;
    """))

In [142]:
# verify
pd.read_sql("SELECT * FROM dim_products_subcategories LIMIT 5;",engine)

,ProductSubCategoryID,ProductCategoryID,ProductSubCategoryName
0,100,1,AA
1,101,1,AB
2,102,1,AC
3,103,2,BA
4,104,2,BB


In [143]:
# Create fact_transactions and load data
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""
        DROP TABLE IF EXISTS fact_transactions;
    """))

    conn.execute(text("""
        CREATE TABLE fact_transactions AS
        SELECT * FROM stg_transactions;
    """))

In [144]:
# verify
pd.read_sql("SELECT * FROM fact_transactions LIMIT 5;", engine)

,TransactionID,AccountID,TransactionDate,TransactionAmount,TransactionType,TransactionChannel,ProductID,Status
0,1,162,2023-02-24,10000.00,Credit,Web,1019,Success
1,2,154,2022-08-31,10000.00,Credit,Web,1001,Success
2,3,179,2020-01-03,4883.80,Debit,Web,1024,Success
3,4,138,2023-11-29,-3911.93,Credit,Atm,1021,Success
4,5,159,2024-02-18,-2211.72,Credit,Mobile,1001,Success


In [145]:
pd.read_sql("SHOW TABLES;", engine)

,Tables_in_fintech_analytics_dw
0,dim_accounts
1,dim_customers
2,dim_customers_usa
3,dim_products
4,dim_products_categories
5,dim_products_subcategories
6,fact_transactions
7,stg_accounts
8,stg_customers
9,stg_customers_usa


# Adding Primary Keys and Foreign Keys

After creating the dimensional warehouse tables, primary keys and foreign key relationships were added to improve:

- data integrity
- table relationships
- query performance
- warehouse consistency
- Power BI relationship modeling

Primary keys uniquely identify each record within a table, while foreign keys establish relationships between dimension tables and the fact table.

These relationships help create a structured star schema model commonly used in business intelligence and analytics environments.

In [146]:
# Add primary keys to dimension and fact tables
with engine.begin() as conn:
    conn.execute(text("""ALTER TABLE dim_accounts 
                      ADD PRIMARY KEY (AccountID)"""))
    
    conn.execute(text(""" ALTER TABLE dim_customers
                      ADD PRIMARY KEY (CustomerID)"""))
    
    conn.execute(text(""" ALTER TABLE dim_customers_usa
                      ADD PRIMARY KEY (CustomerID)""")) 
    
    conn.execute(text(""" ALTER TABLE dim_products
                      ADD PRIMARY KEY (ProductID)"""))
    
    conn.execute(text(""" ALTER TABLE dim_products_categories
                      ADD PRIMARY KEY (ProductCategoryID)"""))  
    
    conn.execute(text(""" ALTER TABLE dim_products_subcategories
                      ADD PRIMARY KEY (ProductSubCategoryID)"""))   
    
    conn.execute(text(""" ALTER TABLE fact_transactions
                      ADD PRIMARY KEY (TransactionID)"""))  
    


In [147]:
# Add foreign keys to fact table verify that all foreign key relationships are valid
tables = [
    "dim_customers",
    "dim_customers_usa",
    "dim_accounts",
    "dim_products",
    "dim_products_categories",
    "dim_products_subcategories",
    "fact_transactions"
]

for table in tables:
    print(f"\n---{table}---")
    display(pd.read_sql(f"SHOW INDEX FROM {table};", engine))


---dim_customers---


,Table,Non_unique,Key_name,Seq_in_index,Column_name,Collation,Cardinality,Sub_part,Packed,Null,Index_type,Comment,Index_comment,Visible,Expression
0,dim_customers,0,PRIMARY,1,CustomerID,A,100,None,None,,BTREE,,,YES,None



---dim_customers_usa---


,Table,Non_unique,Key_name,Seq_in_index,Column_name,Collation,Cardinality,Sub_part,Packed,Null,Index_type,Comment,Index_comment,Visible,Expression
0,dim_customers_usa,0,PRIMARY,1,CustomerID,A,100,None,None,,BTREE,,,YES,None



---dim_accounts---


,Table,Non_unique,Key_name,Seq_in_index,Column_name,Collation,Cardinality,Sub_part,Packed,Null,Index_type,Comment,Index_comment,Visible,Expression
0,dim_accounts,0,PRIMARY,1,AccountID,A,193,None,None,,BTREE,,,YES,None



---dim_products---


,Table,Non_unique,Key_name,Seq_in_index,Column_name,Collation,Cardinality,Sub_part,Packed,Null,Index_type,Comment,Index_comment,Visible,Expression
0,dim_products,0,PRIMARY,1,ProductID,A,27,None,None,,BTREE,,,YES,None



---dim_products_categories---


,Table,Non_unique,Key_name,Seq_in_index,Column_name,Collation,Cardinality,Sub_part,Packed,Null,Index_type,Comment,Index_comment,Visible,Expression
0,dim_products_categories,0,PRIMARY,1,ProductCategoryID,A,3,None,None,,BTREE,,,YES,None



---dim_products_subcategories---


,Table,Non_unique,Key_name,Seq_in_index,Column_name,Collation,Cardinality,Sub_part,Packed,Null,Index_type,Comment,Index_comment,Visible,Expression
0,dim_products_subcategories,0,PRIMARY,1,ProductSubCategoryID,A,9,None,None,,BTREE,,,YES,None



---fact_transactions---


,Table,Non_unique,Key_name,Seq_in_index,Column_name,Collation,Cardinality,Sub_part,Packed,Null,Index_type,Comment,Index_comment,Visible,Expression
0,fact_transactions,0,PRIMARY,1,TransactionID,A,10095,None,None,,BTREE,,,YES,None


In [148]:
# Add foreign key dim_accounts.CustomerID -> dim_customers.CustomerID
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""
        ALTER TABLE dim_accounts
        ADD CONSTRAINT fk_customer
        FOREIGN KEY (CustomerID)
        REFERENCES dim_customers(CustomerID);
    """))

In [149]:
# Add foreign key fact_transactions.AccountID -> dim_accounts.AccountID
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""
        ALTER TABLE fact_transactions
        ADD CONSTRAINT fk_account
        FOREIGN KEY (AccountID)
        REFERENCES dim_accounts(AccountID);
    """))

In [150]:
# Add foreign key fact_transactions.AccountID -> dim_accounts.AccountID
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""
        ALTER TABLE fact_transactions
        ADD CONSTRAINT fk_product
        FOREIGN KEY (ProductID)
        REFERENCES dim_products(ProductID);
    """))

In [151]:
# Add foreign key dim_products.ProductCategoryID -> dim_products_categories.ProductCategoryID
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""
        ALTER TABLE dim_products
        ADD CONSTRAINT fk_subcategory
        FOREIGN KEY (ProductSubcategoryID)
        REFERENCES dim_products_subcategories(ProductSubCategoryID);
    """))

In [152]:
# Add foreign key dim_products.ProductCategoryID -> dim_products_categories.ProductCategoryID
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""
        ALTER TABLE dim_products_subcategories
        ADD CONSTRAINT fk_category
        FOREIGN KEY (ProductCategoryID)
        REFERENCES dim_products_categories(ProductCategoryID);
    """))

In [153]:
# verify foreign key relationships
pd.read_sql("""
SELECT
    TABLE_NAME,
    COLUMN_NAME,
    CONSTRAINT_NAME,
    REFERENCED_TABLE_NAME,
    REFERENCED_COLUMN_NAME
FROM information_schema.KEY_COLUMN_USAGE
WHERE TABLE_SCHEMA = 'fintech_analytics_dw'
AND REFERENCED_TABLE_NAME IS NOT NULL;
""", engine)

,TABLE_NAME,COLUMN_NAME,CONSTRAINT_NAME,REFERENCED_TABLE_NAME,REFERENCED_COLUMN_NAME
0,dim_accounts,CustomerID,fk_customer,dim_customers,CustomerID
1,dim_products,ProductSubcategoryID,fk_subcategory,dim_products_subcategories,ProductSubCategoryID
2,dim_products_subcategories,ProductCategoryID,fk_category,dim_products_categories,ProductCategoryID
3,fact_transactions,AccountID,fk_account,dim_accounts,AccountID
4,fact_transactions,ProductID,fk_product,dim_products,ProductID


In [154]:
# verify specific foreign key constraint
pd.read_sql("""
SELECT CONSTRAINT_NAME, TABLE_NAME
FROM information_schema.TABLE_CONSTRAINTS
WHERE TABLE_SCHEMA = 'fintech_analytics_dw'
AND CONSTRAINT_NAME = 'fk_customer';
""", engine)

,CONSTRAINT_NAME,TABLE_NAME
0,fk_customer,dim_accounts


In [155]:
# calculate customer count by status 
pd.read_sql("""
SELECT
    Status,
    COUNT(*) AS customer_count
FROM dim_customers
GROUP BY Status;
""", engine)

,Status,customer_count
0,Active,33
1,Inactive,29
2,Suspended,38


### Customer Status Analysis

The customer base contains a high number of suspended customers (38), followed by active customers (33) and inactive customers (29).

This may indicate:
- customer compliance issues
- account restrictions
- inactive platform engagement
- operational risk monitoring opportunities

In [ ]:
# calculate transaction count and total amount by transaction type
pd.read_sql("""
SELECT
    TransactionType,
    COUNT(*) AS transaction_count,
    SUM(TransactionAmount) AS total_amount
FROM fact_transactions
GROUP BY TransactionType;
""", engine)

,TransactionType,transaction_count,total_amount
0,Credit,3963,-9898732.50
1,Debit,6037,15123070.06


### Transaction Type Analysis

Debit transactions represented the majority of platform activity, with 6,037 transactions generating a total transaction amount above 15.1 million.

Credit transactions accounted for 3,963 transactions with a negative total amount, reflecting outgoing balances, repayments, or credit-related financial movements within the platform.

This analysis helps monitor:
- transaction behavior
- payment activity
- financial movement trends
- platform transaction distribution

Negative credit transaction amounts may represent outgoing financial movements such as repayments, withdrawals, or credit-related deductions within the platform.

In [ ]:
# calculate average transaction amount by transaction channel 
pd.read_sql("""
SELECT
    c.Region,
    COUNT(t.TransactionID) AS transaction_count,
    ROUND(SUM(t.TransactionAmount), 2) AS total_transaction_amount
FROM dim_customers c
JOIN dim_accounts a
    ON c.CustomerID = a.CustomerID
JOIN fact_transactions t
    ON a.AccountID = t.AccountID
GROUP BY c.Region
ORDER BY total_transaction_amount DESC;
""", engine)

,Region,transaction_count,total_transaction_amount
0,Colorado,1705,1018542.03
1,Kansas,1656,959449.58
2,Massachusetts,1705,789116.79
3,Alaska,1050,610548.35
4,Utah,1280,515438.90
5,California,955,467510.80
6,New Mexico,705,442983.79
7,Florida,944,420747.32


### Regional Transaction Analysis

Colorado generated the highest total transaction amount, exceeding 1 million across 1,705 transactions, followed closely by Kansas and Massachusetts.

The analysis shows strong transaction activity concentrated within a few regions, while Florida and New Mexico generated comparatively lower transaction volumes and transaction values.

This regional analysis can support:
- regional performance monitoring
- customer activity analysis
- operational decision-making
- market performance evaluation

In [159]:
# Product Performance Analysis
pd.read_sql("""
SELECT
    p.ProductName,
    COUNT(t.TransactionID) AS transaction_count,
    ROUND(SUM(t.TransactionAmount), 2) AS total_transaction_amount
FROM fact_transactions t
JOIN dim_products p
    ON t.ProductID = p.ProductID
GROUP BY p.ProductName
ORDER BY total_transaction_amount DESC
LIMIT 10;
""", engine)

,ProductName,transaction_count,total_transaction_amount
0,ACA,365,295503.53
1,BBB,374,272570.45
2,ABA,363,269460.12
3,AAB,372,264352.26
4,BAB,407,256664.64
5,AAC,375,243926.01
6,CAB,370,232875.77
7,CBB,369,231722.33
8,BCB,339,227304.38
9,CAA,367,208210.37


### Product Performance Analysis

Product ACA generated the highest transaction value, exceeding 295K across 365 transactions, followed by BBB and ABA products.

The analysis indicates that a small group of products contributes significantly to total transaction activity and transaction value across the platform.

This type of analysis helps support:
- product performance monitoring
- transaction trend analysis
- KPI reporting
- high-performing product identification

In [160]:
# Transaction Channel Analysis
pd.read_sql("""
SELECT
    TransactionChannel,
    COUNT(*) AS transaction_count,
    ROUND(SUM(TransactionAmount), 2) AS total_transaction_amount
FROM fact_transactions
GROUP BY TransactionChannel
ORDER BY total_transaction_amount DESC;
""", engine)

,TransactionChannel,transaction_count,total_transaction_amount
0,Mobile,5044,2402181.27
1,Web,2990,1658878.21
2,Atm,1966,1163278.08


### Transaction Channel Analysis

Mobile transactions generated the highest transaction activity and transaction value, contributing more than 2.4 million across 5,044 transactions.

Web transactions ranked second, while ATM transactions represented the lowest transaction activity among the available channels.

The analysis indicates strong customer preference toward digital transaction channels, particularly mobile-based financial activity.

In [162]:
# Monthly Transaction Trend Analysis
pd.read_sql("""
SELECT
    DATE_FORMAT(TransactionDate, '%%Y-%%m') AS transaction_month,
    COUNT(*) AS transaction_count,
    ROUND(SUM(TransactionAmount), 2) AS total_transaction_amount
FROM fact_transactions
GROUP BY transaction_month
ORDER BY transaction_month;
""", engine)

,transaction_month,transaction_count,total_transaction_amount
0,2020-01,127,87178.16
1,2020-02,118,30442.85
2,2020-03,157,85508.24
3,2020-04,129,-9066.10
4,2020-05,124,6027.76
...,...,...,...
67,2025-08,134,77615.65
68,2025-09,151,105913.47
69,2025-10,135,162965.93
70,2025-11,160,93425.85


### Monthly Transaction Trend Analysis

Monthly transaction activity was analyzed from 2020 to 2025 to understand transaction volume and value trends over time.

The results show that monthly transaction counts remain relatively stable across the period, while total transaction values fluctuate between positive and negative amounts depending on credit and debit movement.

This analysis is useful for:
- monthly KPI tracking
- transaction trend monitoring
- Power BI line chart development
- identifying high and low transaction-value periods

### Technical Note

Double percentage symbols (`%%`) were used inside the `DATE_FORMAT()` function because `%` is treated as a special formatting character within `pd.read_sql()` queries.

Example:

DATE_FORMAT(TransactionDate, '%%Y-%%m')

This ensures the correct MySQL date formatting syntax (`%Y-%m`) is passed properly from Python to MySQL.

In [163]:
# Yearly Transaction Trend Analysis
pd.read_sql("""
SELECT
    YEAR(TransactionDate) AS transaction_year,
    COUNT(*) AS transaction_count,
    ROUND(SUM(TransactionAmount), 2) AS total_transaction_amount
FROM fact_transactions
GROUP BY transaction_year
ORDER BY transaction_year;
""", engine)

,transaction_year,transaction_count,total_transaction_amount
0,2020,1684,701542.14
1,2021,1681,887031.78
2,2022,1649,878532.74
3,2023,1654,914758.54
4,2024,1650,618482.00
5,2025,1682,1223990.36


### Yearly Transaction Trend Analysis

Transaction activity remained relatively stable between 2020 and 2025, with yearly transaction counts consistently averaging around 1,650 transactions.

Total transaction value showed gradual growth over time, with 2025 generating the highest transaction amount, exceeding 1.2 million.

The yearly trend analysis supports:
- long-term KPI monitoring
- yearly performance evaluation
- executive reporting
- trend visualization in Power BI dashboards

In [165]:
# Product Category Performance Analysis
pd.read_sql("""
SELECT
    pc.ProductCategoryName,
    COUNT(t.TransactionID) AS transaction_count,
    ROUND(SUM(t.TransactionAmount), 2) AS total_transaction_amount
FROM fact_transactions t
JOIN dim_products p
    ON t.ProductID = p.ProductID
JOIN dim_products_subcategories ps
    ON p.ProductSubcategoryID = ps.ProductSubCategoryID
JOIN dim_products_categories pc
    ON ps.ProductCategoryID = pc.ProductCategoryID
GROUP BY pc.ProductCategoryName
ORDER BY total_transaction_amount DESC;
""", engine)

,ProductCategoryName,transaction_count,total_transaction_amount
0,A,3247,1824773.18
1,C,3336,1717051.33
2,B,3417,1682513.05


### Product Category Performance Analysis

Product Category A generated the highest transaction value, exceeding 1.82 million, followed closely by Categories C and B.

Although transaction counts were relatively similar across all categories, Category A produced stronger overall transaction value, indicating higher-value financial activity within that category.

This analysis supports:
- category-level KPI reporting
- product strategy analysis
- transaction value monitoring
- business performance evaluation

In [166]:
# Customer Segmentation Analysis
pd.read_sql("""
SELECT
    Region,
    Status,
    COUNT(*) AS customer_count
FROM dim_customers
GROUP BY Region, Status
ORDER BY Region, customer_count DESC;
""", engine)

,Region,Status,customer_count
0,Alaska,Active,4
1,Alaska,Suspended,3
2,Alaska,Inactive,3
3,California,Suspended,5
4,California,Inactive,3
5,California,Active,2
6,Colorado,Suspended,7
7,Colorado,Inactive,5
8,Colorado,Active,4
9,Florida,Suspended,4


### Customer Segmentation Analysis

Customer distribution varied across regions and customer statuses, with several regions showing a higher number of suspended customers compared to active customers.

Massachusetts and Colorado recorded the highest number of suspended customers, while Kansas showed stronger active customer representation.

This segmentation analysis helps support:
- regional customer monitoring
- operational risk analysis
- customer engagement tracking
- business reporting and segmentation strategies

In [167]:
# Average Transaction Value by Channel
pd.read_sql("""
SELECT
    TransactionChannel,
    COUNT(*) AS transaction_count,
    ROUND(AVG(TransactionAmount), 2) AS avg_transaction_amount
FROM fact_transactions
GROUP BY TransactionChannel
ORDER BY avg_transaction_amount DESC;
""", engine)

,TransactionChannel,transaction_count,avg_transaction_amount
0,Atm,1966,591.70
1,Web,2990,554.81
2,Mobile,5044,476.25


### Average Transaction Value by Channel

ATM transactions generated the highest average transaction value at approximately 591.70 per transaction, followed by Web transactions.

Although Mobile transactions recorded the highest transaction volume overall, the average transaction value was comparatively lower.

This analysis helps support:
- customer transaction behavior analysis
- channel performance evaluation
- KPI monitoring
- digital banking usage insights

In [168]:
# Top Customers by Transaction Value
pd.read_sql("""
SELECT
    c.CustomerID,
    c.Region,
    c.Status,
    COUNT(t.TransactionID) AS transaction_count,
    ROUND(SUM(t.TransactionAmount), 2) AS total_transaction_amount
FROM dim_customers c
JOIN dim_accounts a
    ON c.CustomerID = a.CustomerID
JOIN fact_transactions t
    ON a.AccountID = t.AccountID
GROUP BY
    c.CustomerID,
    c.Region,
    c.Status
ORDER BY total_transaction_amount DESC
LIMIT 10;
""", engine)

,CustomerID,Region,Status,transaction_count,total_transaction_amount
0,16,Colorado,Active,166,155082.64
1,27,Alaska,Active,171,130985.24
2,58,Alaska,Active,180,130853.74
3,41,Colorado,Inactive,156,130018.06
4,93,California,Suspended,128,129079.24
5,95,New Mexico,Inactive,182,127097.25
6,51,Massachusetts,Suspended,97,122051.74
7,50,Massachusetts,Suspended,109,115619.81
8,98,Kansas,Suspended,142,112549.31
9,30,New Mexico,Suspended,159,112538.40


### Top Customers by Transaction Value

The analysis identified several high-value customers generating strong transaction activity across multiple regions.

Customer 16 from Colorado generated the highest total transaction value, exceeding 155K across 166 transactions. Multiple high-performing customers were also identified in Alaska, Massachusetts, and New Mexico.

This type of analysis is valuable for:
- VIP customer identification
- customer value segmentation
- revenue concentration analysis
- stakeholder and executive reporting

In [169]:
# Regional Channel Performance Analysis
pd.read_sql("""
SELECT
    c.Region,
    t.TransactionChannel,
    COUNT(t.TransactionID) AS transaction_count,
    ROUND(SUM(t.TransactionAmount), 2) AS total_transaction_amount
FROM fact_transactions t
JOIN dim_accounts a
    ON t.AccountID = a.AccountID
JOIN dim_customers c
    ON a.CustomerID = c.CustomerID
GROUP BY
    c.Region,
    t.TransactionChannel
ORDER BY total_transaction_amount DESC;
""", engine)

,Region,TransactionChannel,transaction_count,total_transaction_amount
0,Colorado,Mobile,834,486151.53
1,Kansas,Mobile,793,460857.22
2,Colorado,Web,527,340797.86
3,Massachusetts,Mobile,851,310697.43
4,Kansas,Web,507,301235.65
5,Utah,Mobile,683,274108.67
6,California,Mobile,484,268004.68
7,Massachusetts,Web,500,257513.55
8,New Mexico,Mobile,360,238363.39
9,Alaska,Mobile,544,236572.66


### Regional Channel Performance Analysis

Mobile transactions generated the strongest transaction activity and transaction value across most regions, particularly in Colorado, Kansas, and Massachusetts.

Colorado Mobile transactions produced the highest overall transaction value, exceeding 486K, followed closely by Kansas Mobile transactions.

The analysis indicates strong regional adoption of digital transaction channels, especially mobile-based financial activity.

This type of analysis supports:
- regional KPI monitoring
- channel performance evaluation
- customer behavior analysis
- Power BI dashboard reporting
- executive-level business insights

In [ ]:
# SQL Views for Power BI
from sqlalchemy import text

with engine.begin() as conn:

    conn.execute(text("""

        CREATE VIEW vw_transaction_summary AS
        SELECT
            TransactionChannel,
            COUNT(*) AS transaction_count,
            ROUND(SUM(TransactionAmount), 2) AS total_transaction_amount
        FROM fact_transactions
        GROUP BY TransactionChannel;

    """))

In [ ]:
# verify view
pd.read_sql("SELECT * FROM vw_transaction_summary;", engine)

,TransactionChannel,transaction_count,total_transaction_amount
0,Web,2990,1658878.21
1,Atm,1966,1163278.08
2,Mobile,5044,2402181.27


### Power BI Reporting View

A SQL reporting view (`vw_transaction_summary`) was created to simplify Power BI integration and support KPI-based dashboard reporting.

The view provides aggregated transaction metrics by transaction channel, allowing efficient reporting and visualization workflows within Power BI dashboards.

# Project Summary

This project implemented an end-to-end fintech analytics warehouse using Python, Pandas, MySQL, and SQL analytics workflows.

The project included:
- raw data ingestion and cleaning using Pandas
- staging and warehouse table creation
- dimensional modeling using fact and dimension tables
- primary and foreign key relationship management
- SQL-based KPI and business analytics reporting
- customer, product, transaction, and regional analysis
- trend and segmentation analytics
- Power BI reporting view creation for dashboard integration

The final warehouse structure supports scalable fintech reporting and business intelligence dashboard development.